addons/product/models/product_template.py

```diff
@@ -539,42 +558,68 @@ def _get_valid_product_template_attribute_lines(self):
         value to input), indeed single value attributes can be used to filter
         products among others based on that attribute/value.
 
-        This method is necessary because it was previously possible to save a
-        line without any value on it, so the database might not be consistent in
-        that regard.
+        A product attribute value is considered valid for a template if it is
+        defined on a product template attribute line.
 
-        :return: all the valid product template attribute lines of this template
-        :rtype: recordset `product.template.attribute.line`
+        A product attribute is considered valid for a template if it
+        has at least one possible value set on the template.
+
+        For what is considered an archived variant, see `_has_valid_attributes`.
         """
-        self.ensure_one()
-        return self.attribute_line_ids.filtered(lambda ptal: ptal.value_ids)
+        archived_variants = self.env['product.product'].search([('product_tmpl_id', 'in', self.ids), ('active', '=', False)])
+
+        # prefetch
+        self.mapped('attribute_line_ids.value_ids.id')
```

Comment:
can you confirm your 

(14:34:33) seb@odoo.com: il y a peut être une autre solution que le map pour faire un prefetch si le map est trop lent? j'ai pas besoin de résultat du map là
(14:35:06) seb@odoo.com: genre self.attribute_line_ids[:1].value_ids.id 
(14:35:08) seb@odoo.com: à tester

Reply:
(16:26:08) seb@odoo.com: il sera de toute facon genre 10 et pas 1000 ça c'est sur